include/s4.h

```diff
@@ -0,0 +1,325 @@
+  /// contactable sites, the return code of this function only depends on
+  /// whether the subscriber was added to the local site though.
+  ///
+  /// @param id[in]  - The ID of the subscriber. This must be the default
+  ///                  public identity.
+  /// @param aor[in] - The AoR object to update the subscriber with.
+  /// @param trail   - The SAS trail ID
+  ///
+  /// @return Whether the data has been deleted. This can be one of:
+  ///   OK - The AoR was successfully deleted from at least the local site
+  ///       (or wasn't present in the first place).
+  ///   PRECONDITION_FAILED - The subscriber already exists.
+  ///   SERVER_ERROR - We failed to contact the local store; the subscriber
+  ///                  information is unknown.
+  virtual HTTPCode handle_put(const std::string& id,
+                              const AoR& aor,
+                              SAS::TrailId trail);
+
+  /// This updates a subscriber in the deployment. This only succeeds if the
+  /// subscriber already exists. The subscriber is updated in all contactable
+  /// sites, the return code of this function only depends on whether the
+  /// subscriber was updated in the local site though.
+  ///
+  /// @param id[in]   - The ID of the subscriber. This must be the default
+  ///                   public identity.
+  /// @param po[in]   - The patch object to patch and update the subscriber
+  ///                   with.
+  /// @param aor[out] - The retrieved data. This is only valid if the return
+  ///                   code is HTTP_OK. The caller must delete the AoR.
+  /// @param trail    - The SAS trail ID
+  ///
+  /// @return Whether the data has been deleted. This can be one of:
+  ///   OK - The AoR was successfully deleted from at least the local site
+  ///       (or wasn't present in the first place).
+  ///   NOT_FOUND - The subscriber doesn't exist, so can't be patched.
+  ///   SERVER_ERROR - We failed to contact the local store; the subscriber
+  ///                  information is unknown.
+  virtual HTTPCode handle_patch(const std::string& id,
+                                const PatchObject& po,
+                                AoR** aor,
+                                SAS::TrailId trail);
+
+  /// This deletes the subscriber from the local site. This should only be
+  /// called from another S4, not a client.
+  ///
+  /// @param id[in]    - The ID of the subscriber. This must be the default
+  ///                    public identity.
+  /// @param trail[in] - The SAS trail ID
+  ///
+  /// @return Whether the data has been deleted. This can be one of:
+  ///   NO_CONTENT - The AoR was successfully deleted from the local site (or
+  ///                wasn't present in the first place).
+  ///   SERVER_ERROR - We failed to contact the local store; the subscriber
+  ///                  information is unknown.
+  virtual void handle_remote_delete(const std::string& id,
```

Comment:
I think so - local and remote deletes have very different behaviour on contention. It could be a single function, but I think that makes it harder to reason about. I've made this private.

Reply:
> I can't make it private though, as it's called the local S4 on the remote S4.

They're both the same `S4` class, right? If so, your statement is not consistent with how C++ classes work (I don't think 😋 )